taxonomy/tests/test_api.py

```diff
@@ -0,0 +1,346 @@
+        add_type_for_vocabulary(self.learning_resource_type1.name,
+                                self.vocabulary.id)
+        self.assertEquals([self.learning_resource_type1],
+                          list(self.vocabulary.learning_resource_types.all()))
+        add_type_for_vocabulary(self.learning_resource_type2.name,
+                                self.vocabulary.id)
+        self.assertEquals([self.learning_resource_type1,
+                           self.learning_resource_type2],
+                          list(
+                              self.vocabulary.learning_resource_types.
+                              order_by('name')))
+
+        # duplicate is ignored
+        add_type_for_vocabulary(self.learning_resource_type1.name,
+                                self.vocabulary.id)
+        self.assertEquals([self.learning_resource_type1,
+                           self.learning_resource_type2],
+                          list(
+                              self.vocabulary.learning_resource_types.
+                              order_by('name')))
+
+    def test_create_term(self):
+        """
+        Test create_term
+        """
+        self.assertEquals([self.term],
+                          list(Term.objects.all()))
+
+        term = create_term(self.vocabulary.id, "label", 9)
+        self.assertEquals([self.term, term],
+                          list(Term.objects.order_by('id')))
+
+        self.assertRaises(Vocabulary.DoesNotExist,
+                          lambda: create_term(456, "label", 9))
+        self.assertRaises(IntegrityError,
+                          lambda: create_term(self.vocabulary.id, None, 9))
+        create_term(self.vocabulary.id, "label2", 99.7)
+        # term weight is truncated after being stored in database
+        self.assertEquals(1, Term.objects.filter(weight=99).count())
+
+    def test_create_vocabulary(self):
+        """
+        Test create_vocabulary
+        """
+        self.assertEquals([self.vocabulary],
+                          list(Vocabulary.objects.all()))
+
+        v = create_vocabulary(self.repository.id,
+                              "vocabulary",
+                              "description",
+                              False,
+                              "f",
+                              9)
+        self.assertEquals([self.vocabulary, v],
+                          list(Vocabulary.objects.order_by('id')))
+        self.assertRaises(Repository.DoesNotExist,
+                          lambda: create_vocabulary(9,
+                                                    "vocabulary",
+                                                    "description",
+                                                    False,
+                                                    "m",
+                                                    9))
+
+        # int is cast to string
+        create_vocabulary(self.repository.id,
+                          9,
```

Comment:
You're right, when dealing with a single model, the ModelForm won't use the API. That suggests that maybe the API should use ModelForms and all their baked-in goodness. I've actually done that in the past and it works out well. But it does cause some overhead when you have Python objects and you need to massage them to look like they arrived via HTTP POST.

I'd say as long as the API uses the `full_clean` method of the ModelForm we should be good, as far as not duplicating validation logic.